aggregator/src/aggregator.rs

```diff
@@ -3348,15 +3362,31 @@ impl VdafOps {
                         report_count,
                         checksum,
                     );
-                    try_join!(
-                        tx.put_aggregate_share_job(&aggregate_share_job),
-                        try_join_all(batch_aggregations.into_iter().map(|ba| async move {
-                            tx.update_batch_aggregation(&ba.scrubbed()).await
-                        })),
-                        try_join_all(empty_batch_aggregations.into_iter().map(|ba| async move {
-                            tx.put_batch_aggregation(&ba.scrubbed()).await
-                        }))
-                    )?;
+
+                    tx.put_aggregate_share_job(&aggregate_share_job).await?;
+
+                    // Rather than awaiting all the futures at once, which can cause heap growth
+                    // proportional to the number of batch aggregations to write, put them into an
+                    // buffered stream so that they get run in groups of bounded size. #3857
+                    futures::stream::iter(
+                        batch_aggregations
+                            .into_iter()
+                            // existing batch aggregatons must be updated so zip with true
+                            .zip(repeat(true))
+                            // empty batch aggregations are new so zip with false
+                            .chain(empty_batch_aggregations.into_iter().zip(repeat(false)))
+                            .map(|(ba, is_update)| (ba.scrubbed(), is_update)),
+                    )
+                    .map(Result::Ok)
```

Comment:
We could push this `Ok(...)` into the `map()` above. I suspect that iterators may optimize away better than streams.